From 7341b031c571984d3239d19884faad9964854b07 Mon Sep 17 00:00:00 2001
From: Pavan S <psreenivas@habana.ai>
Date: Wed, 05 Jul 2023 12:03:53 +0300
Subject: [PATCH] [SW-151130] gaudi2: ppboot: add uart init for arc0

There are time constrains for the PCIe link establishment
which is performed by ARC0 firmware. Hence operation such
as uart prints must be avoided. In order to preserve the
debug information, the uart prints are written into a
memory buffer, and flushed to the uart after link is
established.

Signed-off-by: Pavan S <psreenivas@habana.ai>
Change-Id: Ia73683b679323859386de6a183bb2875bc518db0
---

diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index 9d6504d..c156478 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -292,13 +292,13 @@
 	device_init_pll(); /* used for uart and delay */
 }
 
-static void ppboot_time_critical_section_start(void)
+void ppboot_time_critical_section_start(void)
 {
 	/* prints are stored in mem buffer */
 	preboot_uart_rbuf_init();
 }
 
-static void ppboot_time_critical_section_end(void)
+void ppboot_time_critical_section_end(void)
 {
 	preboot_uart_rbuf_flush();
 	preboot_uart_rbuf_disable();
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index e2c7f9e..1208a3b 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -723,6 +723,7 @@
 	 * in PCIe routine which make use of new global data
 	 * structure, should be initialized here before use.
 	 */
+	device_init_uart();
 	device_init_efuse();
 	device_init_pci();
 	device_init_pll();
@@ -732,8 +733,10 @@
 
 void secondary_main(void)
 {
+	ppboot_time_critical_section_start();
 	secondary_device_init();
 	preboot_pcie_monitoring();
+	ppboot_time_critical_section_end();
 
 	/* Put secondary CPU in sleep state until further notice */
 	wfe();
diff --git a/include/preboot.h b/include/preboot.h
index 741ee11..ca4b58b 100644
--- a/include/preboot.h
+++ b/include/preboot.h
@@ -282,6 +282,8 @@
 void preboot_boot_info_init(void);
 void ppboot_boot_info_init(void);
 void preboot_recovery_flag(bool en);
+void ppboot_time_critical_section_start(void);
+void ppboot_time_critical_section_end(void);
 
 #endif /* __ASSEMBLY__ */
 
